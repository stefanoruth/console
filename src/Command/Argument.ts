export class Argument<T = string | undefined> {
	constructor(protected name: string, protected description?: string, protected defaultValue?: T) {}

	/**
	 * Set description.
	 */
	setDescription(value: string) {
		this.description = value

		return this
	}

	/**
	 * Set default value.
	 */
	setDefault(value: T) {
		this.defaultValue = value

		return this
	}
}
