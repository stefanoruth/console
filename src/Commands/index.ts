export * from './HelpCommand'
export * from './ListCommand'
export * from './InspireCommand'
export * from './Command'
export * from './CommandRegistry'
