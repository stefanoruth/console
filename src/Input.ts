export class Input<A = {}, O = {}> {
	protected args: string[]
	// protected arguments: A
	// protected options: O

	constructor() {
		this.args = process.argv.splice(2)
		// this.arguments = {}
	}
}
