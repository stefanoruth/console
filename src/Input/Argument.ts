import { LogicException, InvalidArgumentException } from '../Exceptions'

export enum ArgumentMode {
	required = 1,
	optional = 2,
	isArray = 4,
}

type Mode = keyof typeof ArgumentMode

export class Argument<T = any> {
	protected mode: ArgumentMode

	constructor(
		protected name: string,
		mode?: ArgumentMode | Mode,
		protected description?: string,
		protected defaultValue?: T
	) {
		if (typeof mode === 'undefined') {
			mode = ArgumentMode.optional
		} else if (typeof mode === 'string') {
			mode = ArgumentMode[mode]
		} else if (mode > 7 || mode < 1) {
			throw new InvalidArgumentException(`Argument mode "${mode}" is not valid.`)
		}

		this.mode = mode
	}

	/**
	 * Get name.
	 */
	getName() {
		return this.name
	}

	/**
	 * Returns true if the argument is required.
	 */
	isRequired(): boolean {
		return ArgumentMode.required === (ArgumentMode.required & this.mode)
	}

	/**
	 * Returns true if the argument can take multiple values.
	 */
	isArray(): boolean {
		return ArgumentMode.isArray === (ArgumentMode.isArray & this.mode)
	}

	/**
	 * Sets the default value.
	 */
	setDefault(defaultValue: T) {
		if (ArgumentMode.required === this.mode && defaultValue) {
			throw new LogicException('Cannot set a default value except for ArgumentMode.optional mode.')
		}

		if (this.isArray()) {
			if (!defaultValue) {
				;(defaultValue as any) = []
			} else if (!(defaultValue instanceof Array)) {
				throw new LogicException('A default value for an array argument must be an array.')
			}
		}

		this.defaultValue = defaultValue
	}

	/**
	 * Returns the default value.
	 */
	getDefault(): T | undefined {
		return this.defaultValue
	}

	/**
	 * Returns the description text.
	 */
	getDescription(): string {
		return this.description || ''
	}
}
