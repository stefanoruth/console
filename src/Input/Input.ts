import { Signature } from './Signature'
import { InvalidOptionException, InvalidArgumentException } from '../Exceptions'
import { escapeshellarg, isRunningTestMode } from '../helpers'

export type Args = { [k: string]: any }

export class Input {
	protected tokens: string[] = []
	protected parsed: string[] = []
	protected signature: Signature = new Signature()
	protected options: { [k: string]: string[] | string } = {}
	protected arguments: { [k: string]: string[] | string } = {}
	protected interactive: boolean = true

	constructor(argv?: string[]) {
		if (!argv) {
			argv = isRunningTestMode() ? [] : process.argv.slice(2)
		}

		this.tokens = argv
	}

	/**
	 * Bind a new Signature the input.
	 */
	bind(signature: Signature) {
		this.arguments = {}
		this.options = {}
		this.signature = signature
		this.parse()
	}

	/**
	 * Processes command line arguments.
	 */
	protected parse() {
		let parseOptions = true
		this.parsed = [...this.tokens]
		let token: string | undefined = this.parsed.shift()

		while (token !== undefined) {
			// console.log('parse', token.length, token)

			if (parseOptions && token === '') {
				this.parseArgument(token)
			} else if (parseOptions && token === '--') {
				parseOptions = false
			} else if (parseOptions && token.indexOf('--') === 0) {
				this.parseLongOption(token)
			} else if (parseOptions && token[0] === '-' && token !== '-') {
				this.parseShortOption(token)
			} else {
				this.parseArgument(token)
			}

			token = this.parsed.shift()
		}
	}

	/**
	 * Parses a short option.
	 */
	protected parseShortOption(token: string) {
		const name = token.substr(1)

		if (name.length > 1) {
			if (this.signature.hasShortcut(name[0]) && this.signature.getOptionForShortcut(name[0]).acceptValue()) {
				// an option with a value (with no space)
				this.addShortOption(name[0], name.substr(1))
			} else {
				this.parseShortOptionSet(name)
			}
		} else {
			this.addShortOption(name, null)
		}
	}

	/**
	 * Parses a short option set.
	 */
	protected parseShortOptionSet(name: string) {
		const len = name.length
		// console.log('parseShortOptionSet', name, len)
		for (let i = 0; i < len; i++) {
			if (!this.signature.hasShortcut(name[i])) {
				throw new Error(`The "-${name[i]}" option does not exist.`)
			}
			const option = this.signature.getOptionForShortcut(name[i])
			if (option.acceptValue()) {
				this.addLongOption(option.getName(), i === len - 1 ? null : name.substr(i + 1))
				break
			} else {
				this.addLongOption(option.getName(), null)
			}
		}
	}

	/**
	 * Parses a long option.
	 */
	protected parseLongOption(token: string) {
		const name = token.substr(2)
		const pos = name.indexOf('=')
		// console.log('parseLongOption', token, name, pos)

		if (pos !== -1) {
			const value = name.substr(pos + 1)
			// console.log(value, value.length)
			if (value.length === 0) {
				this.parsed.unshift(value)
			}
			this.addLongOption(name.substr(0, pos), value)
		} else {
			this.addLongOption(name, null)
		}
	}

	/**
	 * Parses an argument.
	 */
	protected parseArgument(token: string) {
		const c = Object.keys(this.arguments).length

		if (this.signature.hasArgument(c)) {
			const arg = this.signature.getArgument(c)
			this.arguments[arg.getName()] = arg.isArray() ? [token] : token
			// if last argument isArray(), append token to last argument
		} else if (this.signature.hasArgument(c - 1) && this.signature.getArgument(c - 1).isArray()) {
			const arg = this.signature.getArgument(c - 1)
			const argValues = this.arguments[arg.getName()]

			if (argValues instanceof Array) {
				argValues.push(token)
			}
			// unexpected argument
		} else {
			const all = this.signature.getArguments()

			if (all.length) {
				throw new Error(`Too many arguments, expected arguments "${all.map(a => a.getName()).join('" "')}".`)
			}

			throw new Error(`No arguments expected, got "${token}".`)
		}
	}

	/**
	 * Adds a short option value.
	 */
	protected addShortOption(shortcut: string, value: any) {
		if (!this.signature.hasShortcut(shortcut)) {
			throw new Error(`The "-${shortcut}" option does not exist.`)
		}
		this.addLongOption(this.signature.getOptionForShortcut(shortcut).getName(), value)
	}

	/**
	 * Adds a long option value.
	 */
	protected addLongOption(name: string, value: any) {
		// console.log('addLongOption', name, value)
		if (!this.signature.hasOption(name)) {
			throw new Error(`The "--${name}" option does not exist.`)
		}
		const option = this.signature.getOption(name)
		// console.log(option)

		if (value !== null && !option.acceptValue()) {
			throw new Error(`The "--${name}" option does not accept a value.`)
		}

		if (['', null].includes(value) && option.acceptValue() && this.parsed.length) {
			// if option accepts an optional or mandatory argument
			// let's see if there is one provided
			const next = this.parsed.shift()

			if (next !== undefined) {
				if ((next[0] && next[0] !== '-') || ['', null].includes(next)) {
					value = next
				} else {
					this.parsed.unshift(next)
				}
			}
		}

		if (value === null) {
			if (option.isValueRequired()) {
				throw new Error(`The "--${name}" option requires a value.`)
			}

			if (!option.isArray() && !option.isValueOptional()) {
				value = true
			}
		}

		if (option.isArray()) {
			if (typeof this.options[name] === 'undefined') {
				this.options[name] = []
			}

			const optionValues = this.options[name]

			if (optionValues instanceof Array) {
				optionValues.push(value)
			}
		} else {
			this.options[name] = value
		}
	}

	/**
	 * Returns the first argument from the raw parameters (not parsed).
	 */
	getFirstArgument(): string | undefined {
		let isOption = false

		for (let i = 0; i < this.tokens.length; i++) {
			const token = this.tokens[i]

			if (token && token[0] === '-') {
				if (token.indexOf('=') !== -1 || typeof this.tokens[i + 1] === 'undefined') {
					continue
				}
				// If it's a long option, consider that everything after "--" is the option name.
				// Otherwise, use the last char (if it's a short option set, only the last one can take a value with space separator)
				let name = token[1] === '-' ? token.substr(2) : token.substr(-1)

				if (typeof this.options[name] === 'undefined' && !this.signature.hasShortcut(name)) {
					// noop
				} else if (
					this.options[name] ||
					(this.options[(name = this.signature.shortcutToName(name))] && this.tokens[i + 1] === this.options[name])
				) {
					isOption = true
				}

				continue
			}

			if (isOption) {
				isOption = false
				continue
			}

			return token
		}
	}

	/**
	 * Returns the value of a raw option (not parsed).
	 *
	 * This method is to be used to introspect the input parameters
	 * before they have been validated. It must be used carefully.
	 * Does not necessarily return the correct result for short options
	 * when multiple flags are combined in the same option.
	 */
	getParameterOption(
		values: string | string[],
		defaultValue: any = false,
		onlyParams: boolean = false
	): string | undefined {
		values = values instanceof Array ? values : [values]

		const tokens = [...this.tokens]

		while (tokens.length > 0) {
			const token = tokens.shift()

			if (onlyParams && token === '--') {
				return defaultValue
			}

			for (const value of values) {
				if (token === value) {
					return tokens.shift()
				}

				// Options with values:
				//   For long options, test for '--option=' at beginning
				//   For short options, test for '-o' at beginning
				const leading = value.indexOf('--') === 0 ? `${value}=` : value

				if (leading !== '' && token && token.indexOf(leading) === 0) {
					return token.substr(leading.length)
				}
			}
		}

		return defaultValue
	}

	/**
	 * Returns a stringified representation of the args passed to the command.
	 */
	toString() {
		return this.tokens
			.map(token => {
				const match = /{^(-[^=]+=)(.+)}/g.exec(token)

				if (match) {
					return match[0] + this.escapeToken(match[1])
				}

				if (token && token[0] !== '-') {
					return this.escapeToken(token)
				}

				return token
			})
			.join(' ')
	}

	/**
	 * Escapes a token through escapeshellarg if it contains unsafe chars.
	 */
	protected escapeToken(token: string): string {
		return /{^[\w-]+$}/.test(token) ? token : escapeshellarg(token)
	}

	/**
	 * Returns true if the raw parameters (not parsed) contain a value.
	 *
	 * This method is to be used to introspect the input parameters
	 * before they have been validated. It must be used carefully.
	 * Does not necessarily return the correct result for short options
	 * when multiple flags are combined in the same option.
	 */
	hasParameterOption(values: string | string[], onlyParams: boolean = false): boolean {
		values = values instanceof Array ? values : [values]

		for (const token of this.tokens) {
			if (onlyParams && '--' === token) {
				return false
			}

			for (const value of values) {
				// Options with values:
				//   For long options, test for '--option=' at beginning
				//   For short options, test for '-o' at beginning
				const leading = value.indexOf('--') === 0 ? value + '=' : value
				if (token === value || ('' !== leading && token.indexOf(leading) === 0)) {
					return true
				}
			}
		}

		return false
	}

	/**
	 * Is this input means interactive?
	 */
	isInteractive() {
		return this.interactive
	}

	/**
	 * Sets the input interactivity.
	 */
	setInteractive(interactive: boolean) {
		this.interactive = interactive
	}

	/**
	 * Returns all the given arguments merged with the default values.
	 */
	getArguments() {
		return {
			...this.signature.getArgumentDefaults(),
			...this.arguments,
		}
	}

	/**
	 * Returns the argument value for a given argument name.
	 */
	getArgument(name: string) {
		if (!this.signature.hasArgument(name)) {
			throw new InvalidArgumentException(`The "${name}" argument does not exist.`)
		}
		return this.arguments[name] || this.signature.getArgument(name).getDefault()
	}

	/**
	 * Returns all the given options merged with the default values.
	 */
	getOptions() {
		return {
			...this.signature.getOptionDefaults(),
			...this.options,
		}
	}

	/**
	 * Returns the option value for a given option name.
	 */
	getOption(name: string) {
		if (!this.signature.hasOption(name)) {
			throw new InvalidOptionException(`The "${name}" option does not exist.`)
		}

		return this.options[name] || this.signature.getOption(name).getDefault()
	}

	/**
	 * Check if an option is parsed in by the user.
	 */
	hasOption(option: string): boolean {
		return this.signature.hasOption(option)
	}

	/**
	 * Check if an argument is parsed in by the user.
	 */
	hasArgument(arg: string): boolean {
		return this.signature.hasArgument(arg)
	}

	/**
	 * List args & options
	 */
	getArgs() {
		return {
			...this.getArguments(),
			...this.getOptions(),
		}
	}
}
