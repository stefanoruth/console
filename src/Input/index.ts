export * from './Argument'
export * from './Input'
export * from './Option'
export * from './Signature'
export * from './ParsedInput'
