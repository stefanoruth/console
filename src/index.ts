export { Application } from './Application'
export { Input, Option, Signature, Argument } from './Input'
export { Output, Terminal, ProgressBar, Table } from './Output'
export { Command } from './Commands'
export * from './Events/EventCollection'
export * from './Exceptions'
