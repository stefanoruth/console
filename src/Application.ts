import { Command, HelpCommand, ListCommand, InspireCommand } from './Commands'
import { Input, Signature, Option, Argument } from './Input'
import { CommandNotFoundException } from './Exceptions'
import { Output, ErrorHandler } from './Output'
import {
	EventDispatcher,
	ApplicationStarting,
	CommandStarting,
	CommandFinished,
	EventListener,
	EventTypes,
} from './Events'

export type Bootstrap = (application: Application) => void

export class Application {
	protected commands: { [key: string]: Command } = {}
	protected wantHelps: boolean = false
	protected runningCommand: Command | null = null
	protected catchExceptions: boolean = true
	protected autoExit: boolean = true
	protected defaultCommand: string = 'list'
	protected singleCommand: boolean = false
	protected help?: string
	protected initialized: boolean = false
	protected events: EventDispatcher = new EventDispatcher()
	protected static bootstrappers: Bootstrap[] = []

	/**
	 * Build Console Application.
	 */
	constructor(protected name?: string, protected version?: string) {
		this.bootstrap()
		this.events.dispatch(new ApplicationStarting())
	}

	/**
	 * Register a new Command.
	 */
	register(commands: Command[]): this {
		commands.forEach((command: Command) => {
			this.addCommand(command)
		})

		return this
	}

	/**
	 * Run the Commands.
	 */
	async run(input?: Input, output?: Output) {
		let exitCode: number = 0

		if (!input) {
			input = new Input()
		}

		if (!output) {
			output = new Output()
		}

		const commandName: string | undefined = this.getCommandName(input)

		this.events.dispatch(new CommandStarting(commandName, input, output))

		try {
			exitCode = await this.doRun(input, output)
		} catch (error) {
			if (!this.catchExceptions) {
				throw error
			}

			this.renderException(error, output)

			exitCode = 1 // error.getCode()
		}

		this.events.dispatch(new CommandFinished(commandName, input, output, exitCode))

		if (this.autoExit) {
			process.exit(exitCode)
		}

		return exitCode
	}

	/**
	 * Runs the current application.
	 * int 0 if everything went fine, or an error code
	 */
	protected async doRun(input: Input, output: Output): Promise<number> {
		let command: Command
		if (true === input.hasParameterOption(['--version', '-V'], true)) {
			output.success(this.getHelp())
			return 0
		}

		try {
			// Makes input.getFirstArgument() able to distinguish an option from an argument.
			input.bind(this.getSignature())
		} catch (e) {
			// Errors must be ignored, full binding/validation happens later when the command is known.
		}

		let name: string | undefined = this.getCommandName(input)

		if (true === input.hasParameterOption(['--help', '-h'], true)) {
			this.wantHelps = true
		}

		if (typeof name === 'undefined') {
			name = this.defaultCommand
		}

		try {
			this.runningCommand = null

			command = this.find(name)
		} catch (error) {
			if (!(error instanceof CommandNotFoundException)) {
				throw error
			}

			console.error(error)

			// Find alternatives

			return 1
		}

		this.runningCommand = command
		const exitCode = await this.doRunCommand(command, input, output)
		this.runningCommand = null

		return exitCode
	}

	/**
	 * Gets the help message.
	 */
	getHelp(): string {
		const name = this.getName()
		if (name) {
			const version = this.getVersion()

			if (version) {
				return `${name} ${version}`
			}
			return name
		}
		return 'Console Tool'
	}

	/**
	 * Returns a registered command by name or alias.
	 */
	protected get(name: string): Command {
		this.init()

		if (!this.has(name)) {
			throw new CommandNotFoundException(`The command "${name}" does not exist.`)
		}

		const command = this.commands[name]

		if (this.wantHelps) {
			this.wantHelps = false
			const helpCommand: HelpCommand = this.get('help') as any
			helpCommand.setCommand(command)
			return helpCommand
		}

		return command
	}

	/**
	 * Returns true if the command exists, false otherwise.
	 */
	protected has(name: string) {
		this.init()

		if (typeof this.commands[name] !== 'undefined') {
			return true
		}

		// ($this -> commandLoader && $this -> commandLoader -> has($name) && $this -> add($this -> commandLoader -> get($name))
		// if () {
		//     return true
		// }

		return false
	}

	/**
	 * Finds a command by name or alias.
	 *
	 * Contrary to get, this command tries to find the best
	 * match if you give it an abbreviation of a name or alias.
	 */
	find(name: string): Command {
		this.init()

		if (typeof this.commands[name] === 'undefined') {
			throw new CommandNotFoundException(name)
		}

		if (this.has(name)) {
			return this.get(name)
		}

		throw new Error('Not yet implmeneted this part.')
	}

	/**
	 * Runs the current command.
	 *
	 * If an event dispatcher has been attached to the application,
	 * events are also dispatched during the life-cycle of the command.
	 */
	protected async doRunCommand(command: Command, input: Input, output: Output) {
		try {
			await command.execute(input, output)
		} catch (error) {
			new ErrorHandler(output).render(error)
			return 1
		}

		return 0
	}

	/**
	 * Renders a caught exception.
	 */
	renderException(e: Error, output: Output) {
		output.newLine()

		new ErrorHandler(output).render(e)

		if (this.runningCommand !== null) {
			output.line(this.runningCommand.getSynopsis() + this.getName())
			output.newLine()
		}
	}

	/**
	 * Gets the name of the application.
	 */
	getName() {
		return this.name
	}

	/**
	 * Gets the version of the application.
	 */
	getVersion() {
		return this.version
	}

	/**
	 * Adds a command object.
	 *
	 * If a command with the same name already exists, it will be overridden.
	 * If the command is not enabled it will not be added.
	 */
	addCommand(command: Command) {
		command.setApplication(this)

		this.commands[command.getName()] = command

		return command
	}

	/**
	 * Gets the default commands that should always be available.
	 */
	protected getDefaultCommands() {
		return [new HelpCommand(), new ListCommand(), new InspireCommand()]
	}

	/**
	 * Fetches a list of commands that are registered
	 */
	getCommands(): Command[] {
		return Object.values(this.commands)
	}

	/**
	 * Add base commands to the list of commands.
	 */
	protected init() {
		if (this.initialized) {
			return
		}

		this.initialized = true

		this.register(this.getDefaultCommands())
	}

	/**
	 * Register a console "starting" bootstrapper.
	 */
	static starting(callback: (application: Application) => void) {
		this.bootstrappers.push(callback)
	}

	/**
	 * Bootstrap the console application.
	 */
	protected bootstrap() {
		Application.bootstrappers.forEach(bootstrapper => {
			bootstrapper(this)
		})
	}

	/**
	 * Gets the name of the command based on input.
	 */
	protected getCommandName(input: Input): string | undefined {
		return this.singleCommand ? this.defaultCommand : input.getFirstArgument()
	}

	/**
	 * Gets the Signature related to this Application.
	 */
	getSignature() {
		return new Signature([
			new Argument('command', undefined, 'The command to execute'),
			new Option('--help', '-h', undefined, 'Display this help message'),
			new Option('--quiet', '-q', undefined, 'Do not output any message'),
			new Option('--version', '-V', undefined, 'Display this application version'),
			// new Option(
			// 	'--verbose',
			// 	'-v|vv|vvv',
			// 	undefined,
			// 	'Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug'
			// ),
			// new Option('--ansi', '', undefined, 'Force ANSI output'),
			// new Option('--no-ansi', '', undefined, 'Disable ANSI output'),
			// new Option('--no-interaction', '-n', undefined, 'Do not ask any interactive question'),
		])
	}

	/**
	 * Listen for a specific event.
	 */
	listen(event: EventTypes, listener: EventListener) {
		this.events.addListener(event, listener)

		return this
	}
}
