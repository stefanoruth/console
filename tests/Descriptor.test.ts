import { Descriptor } from '../src/Output/Descriptor'
import { Argument } from '../src/Input/Argument'
import { Output, Terminal } from '../src/Output'
import { Color, TextStyle } from '../src/Output/Style'
import { Option, Signature, CommandSignature } from '../src/Input'
import { Mock } from 'ts-mockery'
import { Command } from '../src/Commands'
import { Application } from '../src/Application'

class TestDescriptor extends Descriptor {
	public buffer = ''
	public bufferList: string[] = []

	protected write(content: string) {
		this.buffer += content.replace(new RegExp(/\n/g), '#')

		content.split('\n').forEach(text => {
			this.bufferList.push(text)
		})
	}
}

class TestCommand extends Command {
	protected name = 'foobar'

	constructor(definition: CommandSignature = [], description: string = '', help: string = '') {
		super()
		this.application = new Application()
		this.signature = definition
		this.description = description
		this.help = help
	}

	async handle() {
		//
	}
}

const color = Mock.of<Color>({
	apply: (text: string) => {
		return text
	},
})

const output = new Output(new Terminal())
const formatter = new TextStyle(color)

function desc(descObject: any): string {
	return new TestDescriptor(output, formatter).describe(descObject).buffer
}

function desc2(descObject: any): string {
	return new TestDescriptor(output, formatter)
		.describe(descObject)
		.bufferList.filter(line => {
			return line.length
		})
		.join('\n')
}

describe('Descriptor', () => {
	test('Describe unknown type of object', () => {
		expect(() => desc({})).toThrow()
	})

	test('Describe an Argument', () => {
		expect(desc(new Argument('foo'))).toBe('  foo')
		expect(desc(new Argument('foo', 'required'))).toBe('  foo')
		expect(desc(new Argument('foo', 'optional'))).toBe('  foo')
		expect(desc(new Argument('foo', 'isArray'))).toBe('  foo')
		expect(desc(new Argument('foo', 'required', 'Foobar'))).toBe('  foo Foobar')
		expect(desc(new Argument('foo', 'optional', 'Foobar'))).toBe('  foo Foobar')
		expect(desc(new Argument('foo', 'isArray', 'Foobar'))).toBe('  foo Foobar')
		expect(desc(new Argument('foo', 'required', 'Foobar', 'baz'))).toBe('  foo Foobar [default: "baz"]')
		expect(desc(new Argument('foo', 'optional', 'Foobar', 'baz'))).toBe('  foo Foobar [default: "baz"]')
		expect(desc(new Argument('foo', 'isArray', 'Foobar', 'baz'))).toBe('  foo Foobar [default: "baz"]')
	})

	test('Describe an Option', () => {
		expect(desc(new Option('foo'))).toBe('      --foo')
		expect(desc(new Option('foo', 's'))).toBe('  -s, --foo')
		expect(desc(new Option('foo', 's', 'none'))).toBe('  -s, --foo')
		expect(desc(new Option('foo', 's', 'optional'))).toBe('  -s, --foo[=FOO]')
		expect(desc(new Option('foo', 's', 'required'))).toBe('  -s, --foo=FOO')
		// expect(desc(new Option('foo', 's', OptionMode.optional | OptionMode.isArray, undefined, 'baz'))).toBe('  -s, --foo')
	})

	test('Describe an Signature', () => {
		expect(desc(new Signature())).toBe('')
		expect(desc(new Signature([new Argument('foo')]))).toBe('Arguments:#  foo#')
		expect(desc(new Signature([new Option('bar')]))).toBe('Options:#      --bar')
		expect(desc(new Signature([new Argument('foo'), new Option('bar')]))).toBe('Arguments:#  foo##Options:#      --bar')
	})

	test('Describe an Command', () => {
		expect(desc2(new TestCommand())).toMatchSnapshot()
		expect(desc2(new TestCommand([], 'foobarDesc'))).toMatchSnapshot()
		expect(desc2(new TestCommand([], '', 'foobarHelp'))).toMatchSnapshot()
		expect(desc2(new TestCommand([new Argument('foo')]))).toMatchSnapshot()
		expect(desc2(new TestCommand([new Option('foo')]))).toMatchSnapshot()
		expect(desc2(new TestCommand([new Argument('foo'), new Option('foo')]))).toMatchSnapshot()
	})

	test('Describe an Application', () => {
		expect(desc(new Signature())).toBe('')
		expect(desc(new Signature([new Argument('foo')]))).toBe('Arguments:#  foo#')
		expect(desc(new Signature([new Option('bar')]))).toBe('Options:#      --bar')
		expect(desc(new Signature([new Argument('foo'), new Option('bar')]))).toBe('Arguments:#  foo##Options:#      --bar')
	})
})
