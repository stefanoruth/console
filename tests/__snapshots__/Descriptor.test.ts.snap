// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Descriptor Describe an Command 1`] = `
"Usage:
  foobar
Options:
  -h, --help            Display this help message
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --command-dir     Load commands dynamically from folder
  -n, --no-interaction  Do not ask any interactive question
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug"
`;

exports[`Descriptor Describe an Command 2`] = `
"Description:
  foobarDesc
Usage:
  foobar
Options:
  -h, --help            Display this help message
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --command-dir     Load commands dynamically from folder
  -n, --no-interaction  Do not ask any interactive question
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug"
`;

exports[`Descriptor Describe an Command 3`] = `
"Usage:
  foobar
Options:
  -h, --help            Display this help message
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --command-dir     Load commands dynamically from folder
  -n, --no-interaction  Do not ask any interactive question
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug
Help:
  foobarHelp"
`;

exports[`Descriptor Describe an Command 4`] = `
"Usage:
  foobar [\\\\<foo>]
Arguments:
  foo
Options:
  -h, --help            Display this help message
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --command-dir     Load commands dynamically from folder
  -n, --no-interaction  Do not ask any interactive question
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug"
`;

exports[`Descriptor Describe an Command 5`] = `
"Usage:
  foobar [options]
Options:
      --foo
  -h, --help            Display this help message
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --command-dir     Load commands dynamically from folder
  -n, --no-interaction  Do not ask any interactive question
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug"
`;

exports[`Descriptor Describe an Command 6`] = `
"Usage:
  foobar [options] [--] [\\\\<foo>]
Arguments:
  foo
Options:
      --foo
  -h, --help            Display this help message
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --command-dir     Load commands dynamically from folder
  -n, --no-interaction  Do not ask any interactive question
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug"
`;
